### PR TITLE
Nft Overhaul

### LINKED ISSUE #16: Improve NFT wallet
Make it so you can see your entire nft wallet, and maybe discard/sell your nfts

Also the nfts are currently literally fungible because its possible for there to be duplicate addresses

### PR DESCRIPTION
**Pull request by lancemathias**
Fully working NFT bot on the REAL Ethereum blockchain

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -128,8 +128,8 @@ dmypy.json
 # Pyre type checker
 .pyre/
 
-# NFT wallet
-wallet.json
+# NFT wallets
+.db
```
Comment (ericpretzel) [inline .gitignore]:
```suggestion
data/.db
```

maybe these should go in the data folder

File: README.md [hunk 2]
```diff
@@ -27,7 +28,8 @@ Please read [CONTRIBUTING.md](CONTRIBUTING.md) for information on how to contrib
 ## Replication
 
 *Note: wid-bot is intended for small private server use only because it uses the ["message content" privileged intent](https://support-dev.discord.com/hc/en-us/articles/4404772028055-Message-Content-Privileged-Intent-for-Verified-Bots). It will not work as a larger bot (100+ servers) without modification.*
-1. **Create your own [Discord Application](https://discord.com/developers/applications). Make a [Reddit API](https://www.reddit.com/dev/api/) account as well.**
+1. **Create your own [Discord Application](https://discord.com/developers/applications). Make a [Reddit API](https://www.reddit.com/dev/api/) account as well.**  
+For the NFT functionality, a [Moralis](https://moralis.io) API key and [Infura](https://infura.io) gateway token for the Ropsten testnet are required to access the blockchain. Furthemore, the NFT functionality uses real (test) Ethereum, so a Metamask (or similar) wallet with Ropsten Ethereum is required.
```
Comment (ericpretzel) [inline README.md]:
```suggestion
For the NFT functionality, a [Moralis](https://moralis.io) API key and [Infura](https://infura.io) gateway token for the Ropsten testnet are required to access the blockchain. Furthermore, the NFT functionality uses real (test) Ethereum, so a Metamask (or similar) wallet with Ropsten Ethereum is required.  
```

File: config.py [hunk 3]
```diff
@@ -11,11 +12,424 @@
...
-WALLET = os.path.join(DATA_FOLDER, 'wallet.json')
 NFT_SIZE = 600
-PREFIX = 'https://ipfs.io/ipfs/'
-BASE = 'QmWrHFYZarTP2b1qCzG7xB45C7P2f8NvEczUfKDgiv3uEs'
+INFURA_GATEWAY = os.environ['INFURA_GATEWAY']
+BASE_URL = 'https://cdn.discordapp.com/attachments/915513848059027457/945195853725978644/unknown.png'
+CONTRACT_ADDRESS = '0xE551386387B1293a738FB5b0aaB10592d13473A2'
+WID_ADDR = os.environ['WID_ADDR']
+PRIVATE_KEY = os.environ['PRIVATE_KEY']
+ABI = '''[
```
Comment (ericpretzel) [inline config.py]: I think the ABI should go into a separate file because it is quite big

File: config.py [hunk 4]
```diff
@@ -11,11 +12,424 @@
 # folder that all data files are stored in
 DATA_FOLDER = 'data'
 
+# Persistent data storage
+DB_FILE = 'config.db'
```
Comment (ericpretzel) [inline config.py]:
```suggestion
DB_FILE = os.path.join(DATA_FOLDER, 'config.db')
```
Same for in .gitignore, I think the database should go into the data folder

File: config.py (hunk 3 above)
Comment (ericpretzel) [reply config.py]: A good place to load the ABI would probably be in the `setup` function in the NFT cog. Or just read from the file every time you need to access it.

File: extension/nft.py [hunk 5]
```diff
@@ -4,57 +4,125 @@
...
+            await ctx.respond('A skill issue prevented address storage', ephemeral=True)
+            return
+        await ctx.respond('Address successfully recorded!', ephemeral=True)
 
     @nft_group.command(description="Mint an NFT on the Widcoin blockchain.")
-    async def mint(self,
-        ctx: discord.ApplicationContext):
-        await ctx.defer()
+    async def mint(self, ctx: discord.ApplicationContext):
+        global counter
```
Comment (ericpretzel) [inline extension/nft.py]: what is the point of this global variable?

File: extension/nft.py [hunk 6]
```diff
@@ -4,57 +4,125 @@
...
 import discord
+from discord import Option
+import web3
+from web3 import Web3
+import sqlite3
 from discord.ext import commands
-from util.smart_contract import ensure_wallet, mint_nft, nft, show, own_nft
+from util.smart_contract import gen_img, get_address, mint_nft, get_last_nft
+from io import BytesIO
+import requests
```
Comment (ericpretzel) [inline extension/nft.py]: please use an asynchronous http library. I think https://docs.aiohttp.org/en/stable/ is good

File: util/smart_contract.py [hunk 7]
```diff
@@ -3,57 +3,63 @@
...
 from PIL import Image, ImageChops
-from io import BytesIO
 import discord
 import config
-import json
+import sqlite3
+import web3
+from web3 import Web3
 from random import randint
 import requests
```
Comment (ericpretzel) [inline util/smart_contract.py]: async plz

**Review by ericpretzel — CHANGES_REQUESTED** [review]
overall, looks really good lance, thank you for the contribution! I left a few comments that I would like you to look at before we can merge this into the main branch.

**Review by ericpretzel — APPROVED** [review]
you tested these changes right?

**Comment by lancemathias** [comment]
Extensively.

**Pull request merged**